contracts/TreasuryInsurance.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import '@openzeppelin/contracts/utils/Address.sol';
+import '@openzeppelin/contracts/utils/Context.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC721/ERC721.sol';
+import './interfaces/IERC20Allowance.sol';
+import './Treasury.sol';
+
+// Treasury contract with insurance.
+contract TreasuryInsurance is Treasury, ERC721 {
+    mapping(uint256 => uint256) public compLimits;
+    mapping(uint256 => string) public insuranceConditions;
+    mapping(uint256 => uint256) public insuranceCosts;
+    mapping(uint256 => address) public insuranceAssets;
+    mapping(uint256 => uint256) public lastPayTime;
+    mapping(uint256 => uint256) public requestedComps;
+    mapping(uint256 => string) public requestedCases;
+
+    uint256 public tickets;
+
+    event Insured(
+        uint256 id,
+        address insured,
+        address asset,
+        uint256 cost,
+        uint256 compLimit
+    );
+    event Requested(uint256 id, uint256 comp);
+    event PaidInsurance(uint256 id, uint256 payment);
+    event Compensated(uint256 id, address insured, uint256 compLimit);
+
+    constructor(address governance_, address executor_)
+        Treasury(governance_, executor_)
+        ERC721('BITDAO_TREASURY_INSURANCE', 'BITTI')
+    {}
+
+    function insure(
+        address to,
+        address asset,
+        uint256 cost,
+        uint256 compLimit,
+        string memory condition
+    ) external virtual onlyExecutor returns (uint256) {
+        compLimits[tickets] = compLimit;
+        insuranceConditions[tickets] = condition;
+        insuranceCosts[tickets] = cost;
+        insuranceAssets[tickets] = asset;
+        lastPayTime[tickets] = block.timestamp;
+        _safeMint(to, tickets);
+        tickets += 1;
+        emit Insured(tickets, to, asset, cost, compLimit);
+        return tickets;
+    }
+
+    function payInsurance(uint256 id, uint256 period) public virtual {
+        lastPayTime[id] = block.timestamp + period;
+        IERC20(insuranceAssets[id]).transferFrom(
+            _msgSender(),
+            address(this),
+            period * insuranceCosts[id]
+        );
+        emit PaidInsurance(id, period * insuranceCosts[id]);
+    }
+
+    function requestInsurance(
+        uint256 id,
+        uint256 comp,
+        string memory reason
+    ) external virtual {
+        require(
+            ownerOf(id) == _msgSender(),
+            'Insurance::request: requester is owner'
+        );
+        require(
+            comp <= compLimits[id],
+            'Insurance::request: requested comp is less than the limit'
+        );
+        requestedComps[id] = comp;
```

Comment:
question: Are multiple calls for compensation for the same ticket allowed?

If multiple calls are allowed, does `requestedComps[id]` get over-written each time a compensation claim on the same id is made?

Reply:
Yeah, I actually made this one wrong as I should use requestedComps[id] rather than have parameter compensation here, because user requested that and we either grant this exact amount or not and so we make it 0 after that for next possible claims until all limit is used.